### ISSUE TITLE
Custom NN types - "undeclared field 'grad'

### ISSUE THREAD

**Issue by mratsim** [mratsim/Arraymancer#327]
When creating a custom NN type if it doesn't use generics you run into issues with the autograd `backward` method:

For example
```Nim
type ShakespeareNet = object
  # Embedding weight = Encoder
  encoder_w: Variable[Tensor[float32]]

  # GRU RNN = Internal representation
  gru_W3s0, gru_W3sN, gru_U3s, gru_bW3s, bU3s: Variable[Tensor[float32]]

  # Linear layer weight = Decoder
  decoder_w: Variable[Tensor[float32]]
```

Will trigger an error here:

https://github.com/mratsim/Arraymancer/blob/89cd3f4cbac7a9219d404c23915bd06da5d1098c/src/autograd/gates_basic.nim#L30-L31

```
yourfile.nim(71, 22) template/generic instantiation of `backward` from here
../src/autograd/gates_basic.nim(32, 34) Error: undeclared field: 'grad'
```

This happens when the generic types and its method can be materialised early.

The workaround is to use a generic type ...

```Nim
type ShakespeareNet[TT] = object
  # Embedding weight = Encoder
  encoder_w: Variable[TT]

  # GRU RNN = Internal representation
  gru_W3s0, gru_W3sN, gru_U3s, gru_bW3s, bU3s: Variable[TT]

  # Linear layer weight = Decoder
  decoder_w: Variable[TT]
```

Renaming the payload field from variable to something else to avoid collision might also help.
Removing `methods` altogether would also be a solution (https://github.com/mratsim/Arraymancer/issues/302)

**Comment by mratsim** [comment]
Trying to change the following

```Nim
type
  Sgd*[TT] = object
    ## Stochastic gradient descent
    params: seq[Variable[TT]]
    lr: float32 # Learning rate.
```

into

```Nim
type
  Sgd*[T] = object
    ## Stochastic gradient descent
    params: seq[Variable[Tensor[T]]]
    lr: T # Learning rate.
```

also triggers the same issue at the same location.

Changing `payload.variable` field to a name that is not used at all didn't fix the issue:

https://github.com/mratsim/Arraymancer/blob/3ae364dc31647ac580464979d251868a25629ce5/src/autograd/ag_data_structure.nim#L81-L86

**Comment by mratsim** [comment]
fixed by #333
